vignettes/opencl.Rmd

```diff
@@ -0,0 +1,143 @@
+
+The actual speedup for a model will depend on whether the `lpdf/lpmf` functions
+are the bottlenecks of the model and the `lpdf/lpmf` function used. 
+The more computationally complex the function is, the larger the expected speedup.
+The biggest speedups are expected when using the GLM functions.
+
+Use [profiling](../profiling.Rmd) in order to establish the bottlenecks in your model.
+
+## OpenCL runtime
+
+OpenCL is supported on most modern CPUs and GPUs. In order to use
+OpenCL in CmdStanR, an OpenCL runtime for the target device must be installed.
+A guide for the most common devices is available [here](https://mc-stan.org/docs/2_26/cmdstan-guide/parallelization.html#opencl).
+
+## Compiling a model with OpenCL
+
+By default, models in CmdStanR are compiled without OpenCL support. Once OpenCL
+support is enabled, a CmdStan model will make use of OpenCL if the functions
+in the model support it. Technically no changes to a model are required to
+support OpenCL since the choice of using OpenCL is handled by the compiler,
+but it can still be useful to rewrite a model to be more OpenCL-friendly by
+using vectorization as much as possible.
+
+Consider a simple logistic regression with parameters `alpha` and `beta`,
+covariates `X`, and outcome `y`.
+
+```stan
+data {
+  int<lower=1> k;
+  int<lower=0> n;
+  matrix[n, k] X;
+  int y[n];
+}
+parameters {
+  vector[k] beta;
+  real alpha;
+}
+model {
+  target += std_normal_lpdf(beta);
+  target += std_normal_lpdf(alpha);
+  target += bernoulli_logit_glm_lpmf(y | X, alpha, beta);
+}
+```
+
+Some fake data will be useful to run this model:
+
+```{r fake-data, message=FALSE, results='hide'}
+library(cmdstanr)
+
+# Generate some fake data
+n <- 200000
+k <- 20
+X <- matrix(rnorm(n * k), ncol = k)
+
+y <- 3 * X[,1] - 2 * X[,2] + 1
+p <- runif(n)
+y <- ifelse(p < (1 / (1 + exp(-y))), 1, 0)
+mdata <- list(k = ncol(X), n = nrow(X), y = y, X = X)
+```
+
+In this model, most of the computation will be handled by the
+`bernoulli_logit_glm_lpmf` function. Because this is a supported GPU function,
+it should be possible to accelerate it with OpenCL. Check 
+[here](http://mc-stan.org/math/opencl_support.html) for a list of functions
+with OpenCL support.
+
+To build the model with OpenCL support, add
+`cpp_options = list(stan_opencl = TRUE)` to the model compile.
+
+```{r compile-opencl, message=FALSE, results='hide'}
+# Compile the model with STAN_OPENCL=TRUE
+model_cl <- cmdstan_model("opencl-files/bernoulli_logit_glm.stan",
+                          cpp_options = list(stan_opencl = TRUE))
+```
+
+## Running models with OpenCL
+
+Running models with OpenCL requires specifying the OpenCL platform and device
+on which to run the model (there can be multiple). If the system has one GPU
+and no OpenCL CPU runtime, the platform and device IDs of the GPU are typically
+both `0`, but the `clinfo` tool can be used to figure out for sure what devices
```

Comment:
On mac I had to install `clinfo` first. Should we mention that might be necessary?

Reply:
We can, though that is mentioned in the linked OpenCL installation docs: https://mc-stan.org/docs/2_26/cmdstan-guide/parallelization.html#opencl

Reply:
Ah ok, that's fine then